File: 2085-count-common-words-with-one-occurrence/2085-count-common-words-with-one-occurrence.js
/**
 * @param {string[]} words1
 * @param {string[]} words2
 * @return {number}
 */
var countWords = function (words1, words2) {
    let result = 0;

    // count freq
    // check if freq is = 1 and other is 1
    let freq1 = new Map();
    let freq2 = new Map();

    let i = 0;
    let j = 0;
    while (i < words1.length || j < words2.length) {
        if (i < words1.length) {
            const str = words1[i];
            freq1.set(str, freq1.has(str) ? freq1.get(str) + 1 : 1);
        }
        if (j < words2.length) {
            const str = words2[j];
            freq2.set(str, freq2.has(str) ? freq2.get(str) + 1 : 1);
        }

        i += 1;
        j += 1;
    }

    for (const key of freq1.keys()) {
        if (!freq2.has(key) || freq1.get(key) > 1 || freq2.get(key) > 1) {
            continue;
        }
        result += 1;
    }

    return result;
};

File: 2085-count-common-words-with-one-occurrence/2085-count-common-words-with-one-occurrence.java
class Solution {
    public int countWords(String[] words1, String[] words2) {
        AtomicInteger result = new AtomicInteger();
        Map<String, int[]> map = new HashMap<>();

        int i = 0;
        int j = 0;
        while (i < words1.length || j < words2.length) {
            if (i < words1.length) {
                if (!map.containsKey(words1[i])) {
                    int[] a = {1, 0};
                    map.put(words1[i], a);
                } else {
                    int[] a = map.get(words1[i]);
                    a[0] += 1;
                    map.replace(words1[i], a);
                }
            }
            if (j < words2.length) {
                if (!map.containsKey(words2[j])) {
                    int[] b = {0, 1};
                    map.put(words2[j], b);
                } else {
                    int[] b = map.get(words2[j]);
                    b[1] += 1;
                    map.replace(words2[j], b);
                }
            }

            i++;
            j++;
        }

        map.forEach((s, ints) -> {
            if (ints[0] == 1 && ints[1] == 1) {
                result.addAndGet(1);
            }
        });
        return result.get();
    }
}